### ISSUE TITLE
Is there a common way to distribute software written in Janet?

### ISSUE THREAD

**Issue by ryukinix** [janet-lang/janet#34]
In Common Lisp for instance you can load your own code into your REPL and make a binary dump. After that you can use this dump binary as your own executable software and it's possible distribute in this fashion. This is common in game development for CL enthusiasts.

In Clojure there is the .jar stuff from Java, it's tied to JVM, but at least it's something.

So... There is a way to embed a janet software as a standalone executable, portable at same operating-system wise? Like Golang builds: a executable statically linked.

If don't have, is there any plan to develop something like that?

**Comment by bakpakin** [comment]
Currently, the best way is a source distribution.

**Comment by bakpakin** [comment]
BTW, these images could likely be portable across OSs.

**Comment by ryukinix** [comment]
WOW! This indeed a nice feature! I'll be watching the janet development :)

**Comment by bakpakin** [comment]
As of 3b6ff3c09a6dcbe4d177f0ae07094b1aa2d31ab4, Janet can now precompile modules into images via the `-c` flag.

For example:
```
$ build/janet -c examples/colors.janet clrs.jimage
$ build/janet
janet:0:> (import clrs)
nil
janet:14:> (print (clrs/color :green "hello"))
hello # but in green
nil
janet:50:>
```

**Comment by ryukinix** [comment]
Nice improvement! I'll test it later. Good work!

**Comment by sepisoad** [comment]
keep up the good work

**Comment by andrewchambers** [comment]
I think a git repository that is something like 'janet-lang/pkgs' which just contains janet tables of urls and cryptographic hashes might be nice. Sort of like a registry of import paths.

**Comment by bakpakin** [comment]
> I think a git repository that is something like 'janet-lang/pkgs' which just contains janet tables of urls and cryptographic hashes might be nice. Sort of like a registry of import paths.

We could definitely set something up like that, its probably a good first step towards a package ecosystem. I think each entry should contain links to source by default or janet images, as these are cross platform, as well as lots of metadata.

For packaging native code, I'm still unsure of the best way to do it that's easy and portable. The cook.janet scripts in some of the modules like [JSON](https://github.com/janet-lang/json) work very well for small, self contained projects, but are certainly a leaky abstraction for cross platform libraries that actually need to link to system specific libraries (libuv comes to mind).

**Comment by bakpakin** [comment]
We have been working on the `jpm` tool for building, and hopefully eventually managing dependencies, distributing software, and several other developer facing tasks. As of now, `jpm install` in a directory with a `project.janet` file can be used to add scripts to some PATH (see [mendoza's project.janet](https://github.com/bakpakin/mendoza/blob/master/project.janet) for a simple example). `jpm` will be Janet's main build tool and language-level package manager.

**Comment by ryukinix** [comment]
Congratulations for this awesome work, @bakpakin. I see a nice future for janet and it's great to see this level of care about software distribution, where so many languages failed miserably.

**Comment by sleepyfox** [comment]
Please don't go down the path of npm (central package hosting), but instead normalise linking to independent github/gitlab/... repos for packages. Thanks. Keep up the good work!.

**Comment by bakpakin** [comment]
So jpm actually has a lot of these features already, including installing dependencies from git urls, and building executables! jpm will never do anything centralized because it’s way more work for me and for people who want to create packages. I am going to push a documentation page to Janet-Lang.org soon, but all of this functionality is in the latest master and will be in 1.2.0. installing git dependencies requires having git installed, and building an executable requires a c compiler, but should both work on most supported systems including windows.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by sleepyfox** [comment]
Thanks for the reply, I'd certainly be up for helping with the doco for new starters.

**Comment by bakpakin** [comment]
All of the functionality for deploying Janet code with `jpm` has been released in 1.2.0, and documentation on it can be found [here](https://janet-lang.org/docs/jpm.html). For updates on the docs, file a bug against http://github.com/janet-lang/janet-lang.org.
